### DISCUSSION TITLE
Open-JDK in Windows

### DISCUSSION THREAD

**Discussion by buddhikac96** [microsoft/openjdk#40] [Q&A] [answered]
Will Open-JDK come as an inbuilt tool in Windows in the Future?

**Comment by kaeufler** [comment]
I am using the java jdk for windows home 64 bit 21H2. Without oracles java. It´s wonderful as a jre.

**Comment by karianna** [comment, marked as answer]
@buddhikac96 - In short, there are no plans to do so.

Longer answer.  OpenJDK is a great tool for Java developers but is not a tool that's used by a large percentage of Windows users and therefore not a good candidate to make part of a default installation.

With regards to a default JRE, since Java 11, the recommended way to deploy a Java runtime is to bundle the runtime modules you need with your application. That is, a Java runtime is bundled with the application as opposed to being externally installed.  You can think of this like electron apps where the Javascript electron runtime is bundled with the app.

Question!  I'm assuming you want OpenJDK built in because there's still a barrier to getting and installing a JDK for Windows?  Is there a way you'd prefer to see OpenJDK made available to you outside of bundling it with windows?

**Reply by buddhikac96** [reply]
Thank you for the clear answer. 

Yes, that was my question. Because I teach Java to kids remotely, and one of the main challenges is installing the JDK in kids' laptops. I neither see any use case apart from that. But won't it be interesting to see the  Java version rather seeing **java is not recognized as . . .** message just after installing the Windows in our laptops as in Ubuntu ;)

**Reply by karianna** [reply]
@buddhikac96 - Thanks for the context - that is indeed a very valid and interesting use case! Would Winget be something that would satisfy your needs for this? https://docs.microsoft.com/en-us/windows/package-manager/winget/

**Reply by jschwartzenberg** [reply]
@karianna I am looking at a similar use case, remote teaching. Reasons this works well on Ubuntu:
1. Installation is just a single command.
2. `JAVA_HOME` is set.
3. `PATH` is covered (through "alternatives").
4. The installation location is fixed, there's always an installation below `/usr/lib/jvm` and certainly below `/usr/lib/jvm/default-java`.
5. Applications that rely on a JDK figure out that the system one is the one and only and use it.
6. Ubuntu's behavior regarding the above has been consistent across major releases for a very long period of time, as far as I remember it never significantly changed ever since Ubuntu was first released.

On Windows I observe:
1. Installation is more complex, usually through a GUI, I don't know if all systems would have `winget`.
2. `JAVA_HOME` is not always set.
3. `PATH` is not always set.
4. The installation location is variable and depends on the installer participants ended up using. While the course recommends a particular installer, participants often use a JDK they installed prior.
5. Applications need to be pointed to the JDK.
6. Typically Windows 7 and 10 are used. Performing the installation and corrections is mostly similar for these, but if a solution for this is developed, having it be consistent and supported for a long time would make all the difference.

Regarding the source code of this OpenJDK distribution, I understand that it is not in a public repository yet. Having VCS access and being able to observe individual commits on changes, does help a lot to make me feel more confident when recommending others to install a piece of software, over simply getting a tarball with the sources from a particular build but no insight in the changes. But also having Winget or a similar tool be able to fetch the source code and build dependencies of a particular package as to enable recompilation of a particular package version similar to `apt`/`yum` is useful. I hope at least the VCS can be made available soon. I never looked at Winget, so I don't know about its feature set to help with recompiling its packages.

**Reply by karianna** [reply]
Thanks for the detailed write-up (CC @brunoborges)!  VCS is something we'll be sorting out in the not-too-distant future (nothing nefarious there, just purely time/prioritization for our engineers) :-).

**Reply by jschwartzenberg** [reply]
You're welcome! I hope this helps. I am relatively unfamiliar with the Windows side of things nowadays, but I can elaborate more on my experience if useful and would be glad to see if this could be matched on Windows.
Great news on the VCS! I am looking forward to see that!

I also wanted to add that one reason there are now so many different providers of an OpenJDK distribution seems mainly because Windows didn't have "one of its own" and this left a gap. On other systems typically you end up installing OpenJDK from the repository as that's simply the quickest way to get going. With Microsoft introducing an OpenJDK, I would expect it to be the defacto one for Windows and Windows users would not require any special tricks anymore either. Hopefully all the various efforts in OpenJDK land can shift on improving the builds as their are part of the various operating systems (repositories) instead of having to maintain all that build infra and those repositories independently.

PS: Similarly if Microsoft considers its non-Windows OpenJDK builds better in any way than the builds from RHEL, Ubuntu, Apple, etc. I would encourage seeking to have those improvements either provided through upstream OpenJDK or by contributing to the building strategies as performed by RHEL, Ubuntu, Apple, etc. for the builds from their OS repositories.
